qcodes/instrument/delegate/delegate_instrument.py

```diff
@@ -329,19 +341,44 @@ def _create_and_add_channel(
             try:
                 channel = self.parse_instrument_path(station, input_params)
             except ValueError as v_err:
-                msg = "Unknown channel path. Try: instrument.chXY"
+                msg = "Unknown channel path."
                 raise ValueError(msg) from v_err
 
         elif isinstance(input_params, Mapping) and channel_wrapper is not None:
-            channel = self.parse_instrument_path(station, input_params["channel"])
-            kwargs = dict(kwargs, **input_params)
+            channel = self.parse_instrument_path(
+                station, input_params["channel"]
+            )
+            wrapper_kwargs = dict(**kwargs, **input_params)
 
-            channel = channel_wrapper(channel.parent, param_name, **kwargs)
+            channel = channel_wrapper(
+                parent=channel.parent,
+                name=channel_name,
+                **wrapper_kwargs
+            )
         else:
-            raise ValueError("Unknown input type.")
+            raise ValueError(
+                "Channels can only be created from existing channels, "
+                "or using a wrapper channel class; "
+                f"instead got {input_params!r} inputs with "
+                f"{channel_wrapper!r} channel wrapper."
+            )
 
-        self.add_submodule(param_name, channel)
+        self.add_submodule(channel_name, channel)
 
     def __repr__(self) -> str:
         params = ", ".join(self.parameters.keys())
         return f"DelegateInstrument(name={self.name}, parameters={params})"
+
+
+def _get_channel_wrapper_class(
+    channel_type: str,
+) -> Optional[Type[InstrumentChannel]]:
+    """Get channel class from string specified in yaml."""
+    if channel_type == "None":
```

Comment:
for the sake of simplicity of this function, can it's input type be Optional[str]? :) it will be same number of lines but cleaner from the typing perspective, meaning that if the argument is a string then it's a "üsable" one

Reply:
I had that first but mypy was complaining. The type of input_params.pop("type", None) is not Optional[str] (but Union[str, Mappingp[str, Any]). So I cast the popped value to str and let _get_channel_wrapper_class take just a string. Is there a better solution to fix the typing issue?

Reply:
(calling ``str`` on an object is not a cast, but a true conversion :) cast would be calling `cast` from `typing` )

aaaah, that's for the channel_type_global case...

hmm, first thing i see is to move the Union in type definition of channels argument of _create_and_add_channels from outside to the key of the outer Mapping: ``Union[Mapping[str, Mapping[str, Any]], Mapping[str, str]]`` -> ``Mapping[str, Union[str, Mapping[str, Any]]]``. have you tried that?

and then in the pop you will have to check that the popped value is not a Mapping (because the "type" key should be a string, not a dict), and then, if mypy still does not understand, then use ``cast`` function to cast the resulting value into optional string. i realize that this will be extra lines of code, but it will be a more readable code because the very flexible "channels" argument is parsed as early as possible into specifics, and then the rest of the code, e.g. _get_channel_wrapper_class becomes much simpler (in that it does not need the ``"None"`` workaround)

Reply:
Thanks @astafan8! Your suggestion works :).